### DISCUSSION TITLE
Why does this project use Scala (and a bunch of other languages)?

### DISCUSSION THREAD

**Discussion by Ashvith10** [ValeLang/Vale#624] [General] [answered]
Apologies, this question is more or less towards understanding of why Scala was used in the project. Correct me if I am wrong, but usually a new compiled language starts from either assembly or a previously made language, from where the compiler is bootstrapped, right? Wouldn't it be ideal to use only a single language, like say, C, or C++, or Rust ?

**Comment by paulocoghi** [comment, marked as answer]
_Disclaimer: I'm not a maintainer_

Some languages are easier (or even more natural) to program certain aspects of a compiler.

This is common among many languages on their beginning and, IMHO, the most likely reason Vale's creator chose multiple languages, including Scala and others, is to to reduce "friction" and make easier to have the minimum working features on the language on its initial phase.

It is also the norm to take years (even a decade or more in many popular cases) to a language be considered mature and, until there, everything that could be used to make the journey easier will surely help (and, sometimes, shorten it a little).

**Reply by paulocoghi** [reply]
I understand that my answer is more superficial and certainly an answer from one of the maintainers will be more enlightening.

**Reply by Verdagon** [reply]
Yep, that's exactly right.

I made as much of the compiler in Scala as possible (the simple imperative kotlin-esque kind of Scala, not the overly functional complex kind of Scala), because it's the only language with good memory safety, good ergonomics (like pattern matching), good build times, and good IDE support.

The backend section is in C++ since that's the language that best works with LLVM. C++ also has pretty good IDE support, and the backend is well-designed enough that we haven't had a single memory safety error, which is nice.
